create RANGE sharding algorithm

strongduanmu opened the pull request:
Ref #5280 .

Changes proposed in this pull request:
- create RANGE sharding algorithm.

tristaZero commented:
@strongduanmu 

Hi, I am glad to read your organized and straightforward code. :)
There is a concern about `range.partition.split.value` which is the key point of this algorithm.

Users need to list all endpoints of each range partitions, i.e, 10,20,30,40...
Besides, it generates complicated code handling. Do you think it is better to prefer the partition volume than listing all endpoints? That is, in this case: instead of `range.partition.split.value=10,20,30,40`, maybe we can consider `partition.volume = 10`. I am not sure, so I'd like to listen to your voice. 😀

tristaZero commented:
BTW, instead of `Fixes #5280 .` which makes this issue closed after merging, `Ref #5280` is suggested, as we still welcome any other great ideas. :)

SanmuWangZJU commented on sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/RangeShardingAlgorithm.java:
```diff
@@ -0,0 +1,128 @@
+    @Override
+    public Collection<String> doSharding(final Collection<String> availableTargetNames, final RangeShardingValue<Long> shardingValue) {
+        Preconditions.checkNotNull(properties.get(RANGE_PARTITION_SPLIT_VALUE), "Range sharding algorithm range partition split value cannot be null.");
+        Map<Integer, Range<Long>> partitionRangeMap = getPartitionRangeMap();
```
for line 61 and 73: 
`doSharding()` method will be called while route value, it may be not efficient to build partitionRangeMap in `doSharding` method, I think it can be used as a static field and  built once while initing instance.

strongduanmu replied:
Good suggestion, I will optimize this question.

strongduanmu replied:
> for line 61 and 73:
> `doSharding()` method will be called while route value, it may be not efficient to build partitionRangeMap in `doSharding` method, I think it can be used as a static field and built once while initing instance.

Good suggestion, I will optimize this question.

tristaZero replied:
@SanmuWangZJU Thanks. :)

strongduanmu replied:
> for line 61 and 73:
> `doSharding()` method will be called while route value, it may be not efficient to build partitionRangeMap in `doSharding` method, I think it can be used as a static field and built once while initing instance.

@SanmuWangZJU @tristaZero 
Hi, I have tried to use static field to define `partitionRangeMap`, but I found that it is not suitable. Users may instantiate multiple different range partition rules in the same application instance to correspond to different requirements. Static initialization means same range partition rule. Perhaps the instance field is more suitable.

tristaZero replied:
@strongduanmu Got it. Maybe we need a `init()` interface function to initialize some local members so as to avoid calculating some local values like `partitionRangeMap`

strongduanmu commented:
> @strongduanmu
> 
> Hi, I am glad to read your organized and straightforward code. :)
> There is my concern about `range.partition.split.value` which is the key point of this algorithm.
> 
> Users need to list all endpoints of each range partitions, i.e, 10,20,30,40...
> Besides, it generates complicated code handling. Do you think it is better to prefer the partition volume than listing all endpoints? That is, in this case: instead of `range.partition.split.value=10,20,30,40`, maybe we can consider `partition.volume = 10`. I am not sure, so I'd like to listen to your voice. 😀

@tristaZero Partition volume looks more concise, but it can only split all values into same interval partition except the first and last partition, this way may not meet the scene of custom partition range.

Users use partition volume to split all values, and may encounter some scenarios where the amount of partition data is small. At this time, it may be more desirable to be able to merge multiple partition ranges, that is, custom partition ranges.

Is it better to provide two Range algorithms? One meets the requirements of flexible scenarios, and the other is more suitable for general scenarios.

strongduanmu commented:
> BTW, instead of `Fixes #5280 .` which makes this issue closed after merging, `Ref #5280` is suggested, as we still welcome any other great ideas. :)

Okay, I will modify it.

tristaZero commented on sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/RangeShardingAlgorithm.java:
```diff
@@ -0,0 +1,128 @@
+        return "RANGE";
+    }
+
+    @Override
```
@Setter and @Getter are recommended.

strongduanmu replied:
Thank you, i will modify it.

tristaZero commented on sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/range/RangeShardingAlgorithmTest.java:
```diff
@@ -0,0 +1,66 @@
+        List<String> availableTargetNames = Lists.newArrayList("t_order_0", "t_order_1", "t_order_2", "t_order_3");
+        List<RouteValue> shardingValues = Lists.newArrayList(new ListRouteValue<>("order_id", "t_order", Lists.newArrayList(0L, 1L, 2L, 4L, 7L)));
+        Collection<String> actual = shardingStrategy.doSharding(availableTargetNames, shardingValues, new ConfigurationProperties(new Properties()));
+        assertThat(actual.size(), is(3));
```
It is suggested to assert some of the specific targets, not only `size()`.

strongduanmu replied:
I got it, thanks!

tristaZero commented on sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/RangeShardingAlgorithm.java:
```diff
@@ -0,0 +1,128 @@
+        Preconditions.checkNotNull(properties.get(RANGE_PARTITION_SPLIT_VALUE), "Range sharding algorithm range partition split value cannot be null.");
+        Map<Integer, Range<Long>> partitionRangeMap = getPartitionRangeMap();
+        Collection<String> result = new LinkedHashSet<>(availableTargetNames.size());
+        for (long value = shardingValue.getValueRange().lowerEndpoint(); value <= shardingValue.getValueRange().upperEndpoint(); value++) {
```
It is likely that`value++` is time-consuming if there is a vast range.
Here are some of my ideas.
For instance,the shardings are as follows,
1,2 | 3,4,5 | 6,7,8,9 | 10,11
So the `range.partition.split.value` is `2,5,9,11`
Imaging the query condition is `2 =< x <= 9`, then we just need to calcute 2==2 and 9<11, therefore the results is `0,1,2`.
That is, we can get the critical values from `range.partition.split.value`, so we only need to compare the bottom point and the top point with those critical values.

strongduanmu replied:
> It is likely that`value++` is time-consuming if there is a vast range.
> Here are some of my ideas.
> For instance,the shardings are as follows,
> 1,2 | 3,4,5 | 6,7,8,9 | 10,11
> So the `range.partition.split.value` is `2,5,9,11`
> Imaging the query condition is `2 =< x <= 9`, then we just need to calcute 2==2 and 9<11, therefore the results is `0,1,2`.
> That is, we can get the critical values from `range.partition.split.value`, so we only need to compare the bottom point and the top point with those critical values.

This suggestion is great, and I will optimize it.

tristaZero commented:
@strongduanmu 
I got your point. This RANGE algorithm is flexible for sharding, which is supposed to be the right choice for users, so please review my comment to perfect this PR. :) 
Besides, we can consider another type of RANGE algorithm. 😀

tristaZero commented:
@strongduanmu Got it. Maybe we need a init() interface function to initialize some local members so as to avoid calculating some local values like partitionRangeMap

tristaZero commented on sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/RangeShardingAlgorithm.java:
```diff
@@ -0,0 +1,124 @@
+ */
+public final class RangeShardingAlgorithm implements StandardShardingAlgorithm<Long> {
+
+    private static final String RANGE_PARTITION_SPLIT_VALUE = "range.partition.split.value";
```
Do you think there is any better names for `range.partition.split.value` to make users easy get its meaning. Like `range.partition.upper.points` or `range.partition.boundary.points`?

tristaZero commented:
Hi @strongduanmu It seems this PR is excellent, so I can not wait to merge it, thanks for your effort! However, there are still some questions that remained for us,
1. Are there any better names for range.partition.split.value to make users easily get its meaning. Like range.partition.upper.points or range.partition.boundary.points?
2.

tristaZero commented:
@strongduanmu Hi
@terrymanu gave another option, i.e, `partition.range`. What do you think?

tristaZero commented:
After more consideration, the second one `init()` is not necessary, maybe.
At first, I considered to initialize `partitionRangeMap` in init(). However, that means `partitionRangeMap` will not be a final member. If so, at the multi-thread condition, it is not safe, which is more important than the performance of `doSharding()`.
What's your idea about this one? Wait for your kind reply. 

Trista

strongduanmu commented:
> @strongduanmu Hi
> @terrymanu gave another option, i.e, `partition.range`. What do you think?

The parameter `partition.range` is better and more easier for users to understand!👍

tristaZero commented:
How about `range.partitions`?
@strongduanmu

strongduanmu commented:
> How about `range.partitions`?
> @strongduanmu



> How about `range.partitions`?
> @strongduanmu

Hi, I found that the `partition.seconds` parameter is defined in the DatetimeShardingAlgorithm class,
Can we define it as `partition.ranges` in the RangeShardingAlgorithm class? 
Ranges can remind users that this algorithm supports configuring multiple ranges for sharding.

strongduanmu commented:
> After more consideration, the second one `init()` is not necessary, maybe.
> At first, I considered to initialize `partitionRangeMap` in init(). However, that means `partitionRangeMap` will not be a final member. If so, at the multi-thread condition, it is not safe, which is more important than the performance of `doSharding()`.
> What's your idea about this one? Wait for your kind reply.
> 
> Trista

Hi, @tristaZero , I agree with you that thread safety is more important. We can use another way to improve performance by providing a `useCache` parameter. 
If users configured fewer partitions, the performance should be able to meet the requirements. But if the num of partition ranges is very large, 
such as 100 partitions, users can set `useCache` parameter to `true`, use the cache to improve performance, just like SQLParseResultCache.

tristaZero commented:
@strongduanmu That is exactly what I am thinking, maybe another PR to rename it as `partition.ranges` is expected. :)

strongduanmu commented:
> @strongduanmu That is exactly what I am thinking, maybe another PR to rename it as `partition.ranges` is expected. :)



> @strongduanmu That is exactly what I am thinking, maybe another PR to rename it as `partition.ranges` is expected. :)

I have modified it, help me review it, thanks!
